Provides mappings for shortcuts used in MoveLess Mode

resolritter opened the pull request:
First of all, thank you for the idea on this plugin.

The changes on this PR introduce flexible bindings instead of hardcoded keys for operation, providing a variable through which users can bind their own keys.

The variable is named g:MoveLess#Mappings and doesn't need to be provided fully (it is extended with the defaults).

---

**Another small change at the bottom of the file:** Prevent suggested shortcut <Leader>m from overriding user map.

resolritter committed:
Provides mappings for shortcuts used in MoveLess Mode.

Prevent suggested shortcut <Leader>m from overriding user map.

resolritter committed:
Provide documentation for g:MoveLess#Mappings.

resolritter committed:
Provide more information regarding the defaults.

anschnapp commented:
Thanks a lot for the nice feature!!
It is my first vim plugin and I'm not so familiar with configurable keybindings...

Thanks also for adjusting the documentation of the vim help section. 
Could you please add your adjustments also to the readme.md file? (It's basicly the same content).
I'm looking foward to merge it after the readme is adjusted too.

If you have any ideas for more optimizations i would be happy to hear it ;)

anschnapp commented on doc/move-less.txt:
```diff
@@ -40,3 +40,36 @@ Any other key: Move less mode will directly be ended, but folding will be tempor
+To modify any of those shortcuts, simply set the variable in your .vimrc, like
+the following example:
+
+let g:MoveLess#Mappings =
```
i have tried this example settings. but i get error because of unmapped keys.
i guess it's necessary to define all keys of move less mappings...

if this is true the example should be adjusted to define all keys...

resolritter replied:
That's strange, it still works for me while only providing a piece of the full map... I'm using Neovim and it's a very recent version, maybe that's why?

Anyway I will change the approach to require the full dictionary. It's simpler :+1:

anschnapp commented on doc/move-less.txt:
```diff
@@ -40,3 +40,36 @@ Any other key: Move less mode will directly be ended, but folding will be tempor
+    'UnfoldUpAndDownK1': 'h',
+    'UnfoldUpAndDownK2': 'L',
+    'StopMoveLess': 'p',
+    'UndoMoveLess': "\<Esc>"
```
i don't like the term `undo` in this context. what to you think about calling it `AbortMoveLess` instead?

anschnapp commented on doc/move-less.txt:
```diff
@@ -40,3 +40,36 @@ Any other key: Move less mode will directly be ended, but folding will be tempor
+  \     'UndoMoveLess': "\<Esc>"
+  \ }
+
+Note that you don't need to provide the whole map, only the entries that you want to
```
only set a subset doesn't work in my test.

i have tested your above example definition. when i start move less mode and press `k` after it i get the following error:
```error detected while processing function MoveLessMode:
line   31:
E716: Key not present in Dictionary: FoldUpAndDownK1
E15: Invalid expression: l:result ==? g:MoveLess#Mappings['FoldDown'] || l:result ==? g:MoveLess#Mappings['FoldUp'] || l:result ==? g:MoveLess#Mappings['FoldUpAndDownK1'] || l:result ==? g:MoveLess#Mappings['Unfo
ldUpAndDownK1'] || l:result ==# g:MoveLess#Mappings['StopMoveLess']
E488: Trailing characters:  l:result ==? g:MoveLess#Mappings['UnfoldUpAndDownK1']
E488: Trailing characters:  l:result ==# g:MoveLess#Mappings['StopMoveLess']
line   89:
E716: Key not present in Dictionary: FoldUpAndDownK1
E15: Invalid expression: l:result ==# g:MoveLess#Mappings['FoldUpAndDownK1'] || l:result ==# g:MoveLess#Mappings['FoldUpAndDownK2']
E488: Trailing characters:  l:result ==# g:MoveLess#Mappings['FoldUpAndDownK2']
line   24:
E716: Key not present in Dictionary: FoldUpAndDownK1
E15: Invalid expression: l:result ==? g:MoveLess#Mappings['FoldDown'] || l:result ==? g:MoveLess#Mappings['FoldUp'] || l:result ==? g:MoveLess#Mappings['FoldUpAndDownK1'] || l:result ==? g:MoveLess#Mappings['Unfo
ldUpAndDownK1']
line   31:
E488: Trailing characters:  l:result ==? g:MoveLess#Mappings['UnfoldUpAndDownK1']
E488: Trailing characters:  l:result ==# g:MoveLess#Mappings['StopMoveLess']
line   89:
E488: Trailing characters:  l:result ==# g:MoveLess#Mappings['FoldUpAndDownK2']
```

anschnapp commented on doc/move-less.txt:
```diff
@@ -40,3 +40,36 @@ Any other key: Move less mode will directly be ended, but folding will be tempor
 As long as no jump or walk over the folding was taken the fold remains, if the move-less mode is activated again in such a situation it will just continue the last mode (by still using the actual folding). 
 This behavior could also be used for going into the move less mode again and make a clean abort of the mode by typing <esc>.
 
+MAPPINGS
```
could you please add the same documentation section in the readme.md. that would be nice :)

resolritter replied:
So the text in the README has slightly different language than the one in the documentation. Do you want both to be the same? As in, provide the full documentation through the README too?

If so, please tell me which version you like more of the two and replace it.

resolritter committed:
Adds the mappings information to the README.md.

A bit of rewording for distinction between mappings and dictionary.

resolritter commented:
@anschnapp sorry to take long to get back on this.

As far as ideas go, what about the idea of saving folds and restoring them on the go? I would be interested in something like this to redo/undo foldings.

anschnapp commented:
@reaysawa thanks for the update!
Did you see that I have added yesterday some more comments on your changes.
Mostly just suggestions for some renaming of the key binding. But I have also described one issue I have faced while trying your example custom binding. Could you please take a second look here.

Thanks also for the idea of an undo redo feature. It would be cool if you can describe in more detail inside a feature request ticket. I'm not sure if undo redo may be the right wording. I could think about some history feature of last foldings (a little bit like the jumplist) or just something like a continue of the last move less search.

I really like to know when and how you would like to use just a fearure and how it would make your editing more effective.

Thanks in advance :)

resolritter commented:
I can't find your comments @anschnapp , where are they? I don't see the review in the Github diff, is something wrong?

I'll come up with some use cases for the folding history idea in a separate issue ticket (￣▽￣)ノ

anschnapp commented:
@reaysawa oh I didn't "commit" my review comments. Should now be visible to you.

resolritter commented on README.md:
```diff
@@ -51,6 +51,40 @@ This behavior could also be used for going into the move less mode again and mak
 
 After such an action the folding will be immediately removed. It is only intended to be used for navigation without moving.
 
+## Customizing the navigation shortcuts
```
Do you like this or the one in doc/move-less.txt? @anschnapp

anschnapp replied:
both are good, but i like this one inside the readme.md more ;)
but it seems like it's still described that you could partially overwrite the key set which should not longer be the case if I understood you correctly.

resolritter committed:
Require the whole dictionary in the variable when updating the key mappings.

resolritter committed:
Update the documentation in the README too.

resolritter committed:
Fixing the documentation in both places.

resolritter commented:
hey @anschnapp I believe I have addressed the changes. I've also changed the Up/Down terminology because it's a bit confusing. Instead it now uses Above/Below (the direction you are navigating into).

Feel free to also push changes to this branch if it's not quite there yet for your liking. :+1:

anschnapp commented on plugin/move-less.vim:
```diff
@@ -39,22 +56,22 @@ function! MoveLessMode()
 
     let l:mode = 'initial'
-    while l:result ==? 'j' || l:result ==? 'k' || l:result ==? 'l' || l:result ==? 'h'
+    while l:result ==? g:MoveLess#Mappings['FoldBelow'] || l:result ==? g:MoveLess#Mappings['FoldAbove'] || l:result ==? g:MoveLess#Mappings['FoldAboveAndBelowK1'] || l:result ==? g:MoveLess#Mappings['UnfoldAboveAndBelowK1']
```
because of the static mapping i have checked the character upper/lower case pairs `l/L` and `h/H` only once but incasesensetive.
you have now substituted it with the custom mapping. but now you have to test for all keys separately because if someone choose instead of `h` and `H` -> `h` and `q` for up and down folding then q would quit the move less mode.

that's true for both if expressions where the key are checked (that the first one don't have the stop key is still correct)

resolritter replied:
do you mean putting checks for overlapping keys? now that you have to pass the whole map explicitly and set each key manually, I think there is less of a chance of this happening. I've found that most plugins don't play around misconfiguration.

anschnapp replied:
no i mean you need to check with `==#` which is case insensetive and you need to take all keys into account which the original checking was included.

So the Original line `l:result ==? 'j' || l:result ==? 'k' || l:result ==? 'l' || l:result ==? 'h'`
was true for the keys: `j`,`J`,`k`,`K`,`l`,`L`,`h`,`H`
which now have 8 counterparts at your configuration map.
but now we have to work with case insensitive checks and take every one explicit into account.
Because we don't know if the user will choose lower upper pairs of the same character for an action and the counterpart of the action like it was before.

resolritter replied:
Good catch! Since this check is repeated, I've moved it into a separate function

anschnapp commented on plugin/move-less.vim:
```diff
@@ -117,7 +134,7 @@ function! MoveLessMode()
             exec "normal! z."
             redraw
-        elseif l:result ==# 'h' || l:result ==# 'L'
+        elseif l:result ==# g:MoveLess#Mappings['UnfoldAboveAndBelowK1'] || l:result ==# g:MoveLess#Mappings['FoldAboveAndBelowK2']
```
i guess here is an error, shouldn't it be `UnfoldAboveAndBelowK2` for the second key?

resolritter replied:
Ugh, I think I should have gone to the first map iteration and redone the substitution manually instead of copying it over... I'll double check in the next commit

anschnapp commented:
Hi @reaysawa thanks for the adjustments. I guess i found 2 little bugs. But apart of that the changes are really nice 👍 
I will merge it after you fix the little issues.

resolritter committed:
Move key checking to a function

Fix inconsistency between pairs of keys

resolritter commented:
@anschnapp if I haven't messed up somewhere, I think it's done... :worried:

anschnapp commented:
@reaysawa looks perfect 👍 . Thank you very much 😄
